crates/rust-analyzer/src/config.rs

```diff
@@ -794,8 +786,27 @@ impl std::ops::Deref for Config {
 }
 
 impl Config {
-    pub fn user_config_path(&self) -> &VfsPath {
-        &self.user_config_path
+    /// Path to the root configuration file. This can be seen as a generic way to define what would be `$XDG_CONFIG_HOME/rust-analyzer/rust-analyzer.toml` in Linux.
+    /// This path is equal to:
+    ///
+    /// |Platform | Value                                 | Example                                  |
+    /// | ------- | ------------------------------------- | ---------------------------------------- |
+    /// | Linux   | `$XDG_CONFIG_HOME` or `$HOME`/.config | /home/alice/.config                      |
+    /// | macOS   | `$HOME`/Library/Application Support   | /Users/Alice/Library/Application Support |
+    /// | Windows | `{FOLDERID_RoamingAppData}`           | C:\Users\Alice\AppData\Roaming           |
+    pub fn user_config_path() -> &'static AbsPath {
+        static USER_CONFIG_PATH: LazyLock<AbsPathBuf> = LazyLock::new(|| {
+            let user_config_path = if let Some(path) = env::var_os("__TEST_RA_USER_CONFIG_DIR") {
+                std::path::PathBuf::from(path)
+            } else {
+                dirs::config_dir()
+                    .expect("A config dir is expected to existed on all platforms ra supports.")
```

Comment:
This feels a bit too heavy-handed. I don't think we should crash `~/.config` doesn't exist, it doesn't get created until someone actually saves a file there.

Reply:
Ah, never mind, it doesn't check for existence.

Reply:
The panic message is misleading 😅

Reply:
I'm still not convinced that not having the directory should be a fatal error. r-a can still work fine enough without one. The Rust toolchain might not, but that's not our problem.

Reply:
There's no config dir on WASM, for example.